### DISCUSSION TITLE
Does Aptos plan to add TreeMap as part of built in data structures?

### DISCUSSION THREAD

**Discussion by 0x-j** [aptos-labs/aptos-developer-discussions#11] [Questions] [answered]
Although Aptos encourages devs to use the pattern of storing app state under each user if possible, but Aptos also supports array and map so devs can store app state in a central place, similar to solidity contracts, this makes building on Aptos more approachable. 

Let's say we are building an NFT marketplace, we don't want to rely on the indexer (due to the complexity of having a custom indexer), we store all the NFT listings in a table. We need an easy way to iterate all the listings with pagination, add or remove listings. If we don't have treemap, we either use `vector` or `table`, but vector makes remove O(N), and table doesn't support iteration. With treemap, it supports both operation and for add or remove it's also not bad as it's O(logN).

**Comment by 0x-j** [comment, marked as answer]
Synced offline with smart contract team, order map is on the roadmap! There's a [`IterableTable`](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/move-examples/data_structures/sources/iterable_table.move) in aptos framework, but it's not not mainnet yet.

**Reply by gregnazario** [reply]
[`SmartTable`](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-stdlib/sources/data_structures/smart_table.move) is an iterable table, and would be preferred for these kinds of use cases.

**Reply by 0x-j** [reply]
Hmm can you write a code snippet on how to iterate a smart table in a contract?
